pennylane/transforms/adjoint_metric_tensor.py

```diff
@@ -0,0 +1,267 @@
+            op.inverse = not op.inverse
+    state = device._apply_operation(state, op)
+
+    return state
+
+
+def _group_operations(tape):
+    """Divide all operations of a tape into trainable operations and blocks
+    of untrainable operations after each trainable one."""
+
+    # Extract tape operations list
+    ops = tape.operations
+    # Find the indices of trainable operations in the tape operations list
+    trainables = np.where([qml.operation.is_trainable(op) for op in ops])[0]
+    # Add the indices incremented by one to the trainable indices
+    split_ids = list(chain.from_iterable([idx, idx + 1] for idx in trainables))
+
+    # Split at trainable and incremented indices to get groups after trainable
+    # operations and single trainable operations (in alternating order)
+    all_groups = np.split(ops, split_ids)
+
+    # Collect trainable operations and groups after trainable operations
+    # the first set of non-trainable ops are the ops "after the -1st" trainable op
+    group_after_trainable_op = dict(enumerate(all_groups[::2], start=-1))
+    trainable_operations = list(chain.from_iterable(all_groups[1::2]))
+
+    return trainable_operations, group_after_trainable_op
+
+
+def adjoint_metric_tensor(circuit, device=None, hybrid=True):
+    r"""Implements the adjoint method outlined in
+    `Jones <https://arxiv.org/abs/2011.02991>`__ to compute the metric tensor.
+
+    A forward pass followed by intermediate partial backwards passes are
+    used to evaluate the metric tensor in :math:`\mathcal{O}(p^2)` operations,
+    where :math:`p` is the number of trainable operations, using 4 state
+    vectors.
+
+    .. note::
+        The adjoint metric tensor method has the following restrictions:
+
+        * As it requires knowledge of the statevector and uses private
+          methods of the used device, only ``"default.qubit"`` devices can be used.
+
+        * We assume the circuit to be composed of unitary gates only and rely
+          on the ``generator`` property of the gates to be implemented.
+          Note also that this makes the metric tensor strictly real-valued.
+
+    Args:
+        circuit (.QuantumTape or .QNode): Circuit to compute the metric tensor of
+        device (.Device): Device to use for the adjoint method
+        hybrid (bool): Whether to take classical preprocessing into account. Ignored if
+            ``circuit`` is a tape.
+
+    Returns:
+        array: the metric tensor of the tape with respect to its trainable parameters.
+        Dimensions are ``(len(trainable_params), len(trainable_params))``.
+
+    """
+    if isinstance(circuit, qml.tape.QuantumTape):
+        return _adjoint_metric_tensor_tape(circuit, device)
+    if isinstance(circuit, (qml.QNode, qml.ExpvalCost)):
+        return _adjoint_metric_tensor_qnode(circuit, device, hybrid)
+
+    raise qml.QuantumFunctionError("The passed object is not a QuantumTape or QNode.")
+
+
+def _adjoint_metric_tensor_tape(tape, device):
```

Comment:
not a fully serious suggestion, just an observation! In fact, the same observation is why we moved https://github.com/PennyLaneAI/pennylane/blob/1fd370d0ee2b14c3ddef9af6e5eccbf8784350a6/pennylane/_qubit_device.py#L843 to the device

Reply:
Okay, off it goes :)

Reply:
Huh, I'm wondering now: Wouldn't the suitable place for `adjoint_jacobian` and `adjoint_metric_tensor` be `devices/default_qubit.py` instead of `_qubit_device.py`?

Reply:
don't worry about doing this, my train of thought got away from me :) we will move it in the future if we need to.

Reply:
Regarding your question: yes, probably. The original reason for adding it to `QubitDevice` was the thinking that _all_ external statevector subclasses would inherit it, and get adjoint support for free.

But it doesn't really follow since it utilizes private methods!

Reply:
Haha, alright :)

Reply:
I see. But aren't QPU devices also `QubitDevice`s? :thinking: